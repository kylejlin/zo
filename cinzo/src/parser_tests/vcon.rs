use super::*;

#[test]
fn vcon() {
    let src = r#"
(
vcon

(
    ind

    Type0

    "Nat"

    // Index types
    ()

    (
        (() ())
        ((0) ())
    )
)

// Variant constructor index - THIS MUST BE A NUMBER LITERAL.
0
)"#;
    let tokens = lex(src).unwrap();
    let cst = parse(tokens).unwrap();
    insta::assert_debug_snapshot!(&cst);
}

#[test]
fn vcon_nonliteral_ind() {
    let src = r#"
    (
    vcon
    
    0
    
    0
    )"#;
    let tokens = lex(src).unwrap();
    let err = parse(tokens).unwrap_err();
    insta::assert_debug_snapshot!(&err);
}

#[test]
fn vcon_nonliteral_vcon_index() {
    let src = r#"
(
vcon

(
    ind

    Type0

    "Nat"

    // Index types
    ()

    (
        (() ())
        ((0) ())
    )
)

(2 3)
)"#;
    let tokens = lex(src).unwrap();
    let err = parse(tokens).unwrap_err();
    insta::assert_debug_snapshot!(&err);
}
