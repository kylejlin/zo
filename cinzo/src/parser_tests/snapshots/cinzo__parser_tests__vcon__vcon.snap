---
source: src/parser_tests/vcon.rs
expression: "&cst"
---
Vcon(
    Vcon {
        lparen: ByteIndex(
            1,
        ),
        ind: Ind {
            lparen: ByteIndex(
                9,
            ),
            type_: UniverseLiteral {
                level: 0,
                start: ByteIndex(
                    24,
                ),
            },
            name: StringLiteral {
                value: "Nat",
                span: (
                    ByteIndex(
                        35,
                    ),
                    ByteIndex(
                        40,
                    ),
                ),
            },
            index_types_lparen: ByteIndex(
                65,
            ),
            index_types: Nil,
            index_types_rparen: ByteIndex(
                66,
            ),
            constructor_defs_lparen: ByteIndex(
                73,
            ),
            constructor_defs: Cons(
                Cons(
                    Nil,
                    VariantConstructorDef {
                        lparen: ByteIndex(
                            83,
                        ),
                        param_types_lparen: ByteIndex(
                            84,
                        ),
                        param_types: Nil,
                        param_types_rparen: ByteIndex(
                            85,
                        ),
                        index_args_lparen: ByteIndex(
                            87,
                        ),
                        index_args: Nil,
                        index_args_rparen: ByteIndex(
                            88,
                        ),
                        rparen: ByteIndex(
                            89,
                        ),
                    },
                ),
                VariantConstructorDef {
                    lparen: ByteIndex(
                        99,
                    ),
                    param_types_lparen: ByteIndex(
                        100,
                    ),
                    param_types: Cons(
                        Nil,
                        Deb(
                            NumberLiteral {
                                value: 0,
                                span: (
                                    ByteIndex(
                                        101,
                                    ),
                                    ByteIndex(
                                        102,
                                    ),
                                ),
                            },
                        ),
                    ),
                    param_types_rparen: ByteIndex(
                        102,
                    ),
                    index_args_lparen: ByteIndex(
                        104,
                    ),
                    index_args: Nil,
                    index_args_rparen: ByteIndex(
                        105,
                    ),
                    rparen: ByteIndex(
                        106,
                    ),
                },
            ),
            constructor_defs_rparen: ByteIndex(
                112,
            ),
            rparen: ByteIndex(
                114,
            ),
        },
        vcon_index: NumberLiteral {
            value: 0,
            span: (
                ByteIndex(
                    179,
                ),
                ByteIndex(
                    180,
                ),
            ),
        },
        rparen: ByteIndex(
            181,
        ),
    },
)
